sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs

```diff
@@ -0,0 +1,107 @@
+        private const string ArchApiVersion = "2019-11-01";
+
+        private readonly string _clientId;
+        private readonly Uri _endpoint;
+
+        public static ManagedIdentitySource TryCreate(CredentialPipeline pipeline, string clientId)
+        {
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string imdsEndpoint = EnvironmentVariables.ImdsEndpoint;
+
+            // if BOTH the env vars IDENTITY_ENDPOINT and IMDS_ENDPOINT are set the MsiType is Azure Arc
+            if (string.IsNullOrEmpty(identityEndpoint) || string.IsNullOrEmpty(imdsEndpoint))
+            {
+                return default;
+            }
+
+            Uri endpointUri;
+            try
+            {
+                endpointUri = new Uri(identityEndpoint);
+            }
+            catch (FormatException ex)
+            {
+                throw new AuthenticationFailedException(IdentityEndpointInvalidUriError, ex);
+            }
+
+            return new AzureArcManagedIdentitySource(pipeline, endpointUri, clientId);
+        }
+
+        private AzureArcManagedIdentitySource(CredentialPipeline pipeline, Uri endpoint, string clientId) : base(pipeline)
+        {
+            _endpoint = endpoint;
+            _clientId = clientId;
+        }
+
+        protected override Request CreateRequest(string[] scopes)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+            Request request = Pipeline.HttpPipeline.CreateRequest();
+            request.Method = RequestMethod.Get;
+            request.Headers.Add("Metadata", "true");
+
+            request.Uri.Reset(_endpoint);
+            request.Uri.AppendQuery("api-version", ArchApiVersion);
+
+            request.Uri.AppendQuery("resource", resource);
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                request.Uri.AppendQuery("client_id", _clientId);
+            }
+
+            return request;
+        }
+
+        protected override async ValueTask<AccessToken> HandleResponseAsync(bool async, TokenRequestContext context, Response response, CancellationToken cancellationToken)
+        {
+            if (response.Status == 401)
+            {
+                if (!response.Headers.TryGetValue("WWW-Authenticate", out string challenge))
+                {
+                    throw new AuthenticationFailedException(NoChallengeErrorMessage);
+                }
+
+                var splitChallenge = challenge.Split(new char[] { '=' }, StringSplitOptions.RemoveEmptyEntries);
+
+                if (splitChallenge.Length != 2)
+                {
+                    throw new AuthenticationFailedException(InvalidChallangeErrorMessage);
+                }
+
+                var authHeaderValue = "Basic " + File.ReadAllText(splitChallenge[1]);
```

Comment:
Do we get `AuthenticationFailedException` when this fails?

Reply:
Yes any exceptions resulting from reading the file will be handled by the credential and wrapped in an `AuthenticationFailedException`.